File: examples/dcps/Listener/cs/src/ListenerDataListener.cs
using System;
using System.IO;

using DDS;
using DDS.OpenSplice;

using ListenerData;
using DDSAPIHelper;

namespace ListenerDataSubscriber
{
    class ListenerDataListener : IDataReaderListener
    {
        public GuardCondition guardCond = new GuardCondition();        
        public bool terminated = false;        
        ReturnCode status = ReturnCode.Error;

        /* Type specific DDS entity */
        private MsgDataReader msgDR;

        public MsgDataReader MsgDR
        {
            get { return msgDR; }
            set { msgDR = value; }
        }
	
        #region IDataReaderListener Members

        public void OnDataAvailable(IDataReader entityInterface)
        {
            Msg[] msgList = null ;
            SampleInfo[] infoSeq = null;
        
            status = msgDR.Read(ref msgList, ref infoSeq, Length.Unlimited, SampleStateKind.Any, ViewStateKind.New, InstanceStateKind.Any);
            ErrorHandler.checkStatus(status, "DataReader.Read");

            if (msgList != null && msgList.Length > 0)
            {
                Console.WriteLine("=== [ListenerDataListener::OnDataAvailable] - msgList.Length : {0}", msgList.Length);
                foreach (Msg msg in msgList)
                {
                    Console.WriteLine("    --- Message Received ---");
                    Console.WriteLine("    userId : {0}", msg.userID);
                    Console.WriteLine("    message : \\ {0}",msg.message);
                }
                status = msgDR.ReturnLoan(ref msgList, ref infoSeq);
                ErrorHandler.checkStatus(status, "DataReader.ReturnLoan");
            }
            guardCond.SetTriggerValue(true);
        }

        public void OnLivelinessChanged(IDataReader entityInterface, LivelinessChangedStatus status)
        {
            Console.WriteLine("=== [ListenerDataListener::OnlivelinessChanged] triggered");
        }

        public void OnRequestedDeadlineMissed(IDataReader entityInterface, RequestedDeadlineMissedStatus status)
        {
            Console.WriteLine("=== [ListenerDataListener::OnRequestedDeadlineMissed] : triggered");
            Console.WriteLine("=== [ListenerDataListener::OnRequestedDeadlineMissed] : stopping");
            terminated = true;
            // unblock the waitset in Subscriber main loop
            guardCond.SetTriggerValue(true);
        }

        public void OnRequestedIncompatibleQos(IDataReader entityInterface, RequestedIncompatibleQosStatus status)
        {
            Console.WriteLine("=== [ListenerDataListener::OnRequestedIncompatibleQos] : triggered");            
        }

        public void OnSampleLost(IDataReader entityInterface, SampleLostStatus status)
        {
            Console.WriteLine("=== [ListenerDataListener::OnRequestedIncompatibleQos] : triggered"); 
        }

        public void OnSampleRejected(IDataReader entityInterface, SampleRejectedStatus status)
        {
            Console.WriteLine("=== [ListenerDataListener::OnSampleRejected] : triggered"); 
        }

        public void OnSubscriptionMatched(IDataReader entityInterface, SubscriptionMatchedStatus status)
        {
            Console.WriteLine("=== [ListenerDataListener::OnSubscriptionMatched] : triggered"); 
        }

        #endregion
    }
}

File: examples/dcps/Listener/java5/ListenerDataListener.java
/*
 *                         OpenSplice DDS
 *
 *   This software and documentation are Copyright 2006 to TO_YEAR PrismTech
 *   Limited, its affiliated companies and licensors. All rights reserved.
 *
 *   Licensed under the Apache License, Version 2.0 (the "License");
 *   you may not use this file except in compliance with the License.
 *   You may obtain a copy of the License at
 *
 *       http://www.apache.org/licenses/LICENSE-2.0
 *
 *   Unless required by applicable law or agreed to in writing, software
 *   distributed under the License is distributed on an "AS IS" BASIS,
 *   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *   See the License for the specific language governing permissions and
 *   limitations under the License.
 *
 */
import java.util.ArrayList;
import java.util.List;

import org.omg.dds.core.GuardCondition;
import org.omg.dds.core.event.DataAvailableEvent;
import org.omg.dds.core.event.RequestedDeadlineMissedEvent;
import org.omg.dds.sub.DataReader;
import org.omg.dds.sub.DataReaderAdapter;
import org.omg.dds.sub.Sample;
import org.omg.dds.sub.Subscriber.DataState;
import org.omg.dds.sub.ViewState;

import ListenerData.*;

public class ListenerDataListener extends DataReaderAdapter<Msg> {

	boolean closed = false;
	GuardCondition guardCond = null;

	public ListenerDataListener(GuardCondition myGC) {
	    guardCond = myGC;
    }

	@Override
    public void onDataAvailable(DataAvailableEvent<Msg> status) {
        /* Handle the incoming data here. */
		DataReader<Msg> reader = status.getSource();
		DataState ds = reader.getParent().createDataState();
		ds = ds.withAnySampleState()
		       .with(ViewState.NEW)
		       .withAnyInstanceState();
		List<Sample<Msg>> samples = new ArrayList<Sample<Msg>>();

		reader.select().dataState(ds).read(samples);

		boolean hasValidData = false;
		if (samples.size() > 0) {
			System.out.println("=== [ListenerDataListener.on_data_available] - msgList.length : "+ samples.size());
			for (Sample<Msg> sample : samples) {
			    Msg msg = sample.getData();
				if (msg != null) {
					hasValidData = true;
					System.out.println("    --- message received ---");
					System.out.println("    userID  : " + msg.userID);
					System.out.println("    Message : \"" + msg.message+ "\"");
				}
			}

			if (hasValidData) {
				// unblock the Waitset in Subscriber main loop
			    guardCond.setTriggerValue(true);
			} else {
				System.out.println("=== [ListenerDataListener.on_data_available] ===> hasValidData is false!");
			}
		}
	}

	@Override
	public void onRequestedDeadlineMissed(RequestedDeadlineMissedEvent<Msg> status) {

		System.out.println("=== [ListenerDataListener.on_requested_deadline_missed] : triggered");
		System.out.println("=== [ListenerDataListener.on_requested_deadline_missed] : stopping");
		// unblock the waitset in Subscriber main loop
		guardCond.setTriggerValue(true);
		closed = true;
	}
}
